src/scripts/modules/add_destination.js

```diff
@@ -0,0 +1,80 @@
+import destinationsList from "../destinationList.js";
+
+let form;
+let uploadBtn;
+let radioBtns;
+let radioVisited;
+let radioBucketlist;
+let allInputs;
+let newCity;
+let cityInput;
+let countryInput;
+let photoUrl;
+
+function cacheDom() {
+  form = document.getElementById("add-city-form");
+  uploadBtn = document.getElementById("upload");
+  radioBtns = document.querySelectorAll("input[type=radio]");
+  radioVisited = document.getElementById("rd-visited");
+  radioBucketlist = document.getElementById("rd-bucketlist");
+  newCity = createCity();
+  cityInput = document.getElementById("inputCity").value;
+  countryInput = document.getElementById("inputCountry").value;
+}
+
+function bindEvents() {
+  form.addEventListener("submit", (event) => {
```

Comment:
So, a few things that need to be changed in order for this to work:

- There's no need to add an event list to uploadButton. By clicking a "type="submit", the form is immediately submitted. It's inside it that the needed validation/submit should be done.
- In the DOM remove `onClick="this.form.reset()"`. That does nothing and usually inline events direct int HTML should be avoided.
- To get the values, you need to add a `change` event listener to each form field to listen to its value updates and then store it in `newCity`. The way you did you get the initial value bue it was missing the listener. I'll give u an example with country field:

```js
  countryInput = document.getElementById("inputCountry");

// ...

  countryInput.addEventListener("change", (ev) => {
    updateCity("country", ev.target.value);
  });
  form.addEventListener("submit", (event) => {
    event.preventDefault();
    console.log(newCity); // it will have the country field populated!
  });
 // remove the eventlisternes to uploadBtn

// ...

function updateCity(fieldName, value) {
  console.log("updateCity:", fieldName, value);
  newCity = {
    ...newCity,
    [fieldName]: value,
  };
}
```

Reply:
I had a feeling you were going to say that about onclick 😄 that was supposed to stop the page from refreshing on click but yeah I noticed it doesn't work. Another way to do that? 

> P.S. Why is the form in the bottom of the page? I almost missed it! It could be in the top, before the images :D.

It's just for now until I make it work, but I see your point 😆 I'll redo the whole thing once the main features are done

Reply:
The page does not refresh on submit because you are already used `e.preventDefault()` in the form submit :)

Reply:
Completely forgot about that 😆 so avoid all inline events, no onfocus="this.value=''" either?

Reply:
Yes, no inline events because they are useless, you cannot make them work because of scope, you can't access specific things about the module, in the HTML side. But the main reason is to keep JS logic together, on the JS side.

`...` is called [Spread syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax). It does a "shallow copy" of the current object or array to a new one. 

In this case, the new one is a mutation of the previous one, so we just do the following, because we are mutating `newCity` directly.

```
newCity[fieldName] = value
```

Reply:
I'm trying to do the pubsub module but after reading and watching tutorials I still can't tell what needs to be published or subscribed. My understanding is that destinationsList.js needs to know about add_destination.js and destinations.js needs to know about destinationsList.js, correct?

Reply:
It's okay, it takes some time to get used to this 'mental model' of pubsub. So, what's needs is:
When a city is added, it needs to announce that new addition (pub). On the other end, the module responsible for rendering the list of destinations, needs to know when this addition was made (sub) so that it can update the list rendering!
Does it make sense? Maybe watching more code tutorials about pubsub can help you understand how to use this concept :)

Reply:
So pub requires some data, that data should be the result of the submit event on the form because that's what ties the module together. Does that sound right? If so, I tired a couple ways to get or store that value (commented out) but I'm obviously missing something. The sub seems to work but I'm not sure

Reply:
Don't worry, Keep trying, even if you can't find the solution, push the code and I'll help u with a solution later tonight :)

Reply:
No rush, I already pushed what I had and don't think I'll be getting closer to the solution anyways 😄 Btw, when I try to use the console to test this out I just get an error, what's the catch? https://youtu.be/nQRXi1SVOow?t=586

Reply:
>  I try to use the console to test this out I just get an error

When you have an error, anything, you must tell what error it is exactly, otherwise, it's impossible to help u :)

Reply:
Sorry 😁 It’s always ‘something is undefined’, e.g. events in events.subscribe(...)

Reply:
Did it you try in this project or in another one, from scratch as it shows in the tutorial?

Based on that error, it seems you tried in this one. If so, that won't allow you to use the console and test that code because the code is "scoped" and it's impossible to access it globally. That's one of the characteristics of module bundlers, like Parcel and Webpack.

Reply:
That clears it up, I tried it in this project.

Reply:
That was why then :) Let me know when you are done with the logic + UI to add new cities, so that I can review it and you can merge it.

The next step: Toggle an existing destination `visited` status :D

Reply:
I'm working on figuring out local storage, that sounded like the next thing. Here's what I think needs to happen:
-the inputs are stored on 'submit' 
- they are loaded and displayed on page load using pubsub
Any problems with this?

Reply:
> I'm working on figuring out local storage, that sounded like the next thing.

Yes, you are right, on first sight, it looks like the next thing to do! I recommended you to do the edition first because it will force you do to a minor refactor that will help you with LS (LocalStorage). But it's totally okay if you do LS right now ;) Just let me know when you are done with this part first :)

Reply:
I'll just push what I already did and then comment it out if neccessary, I might have reached that point where refactoring would've been helpful. I'm stuck trying to not overwrite local storage with each new submit, at this point I'm not sure what I even need to achieve- store each new item as a new key/value pair, push to the same array or something else.

So this button is supposed to send a bucketlist item to the visited list? I can't imagine the UI of it, pinterest is all that comes to mind but that's not very helpful 😄

Reply:
> at this point I'm not sure what I even need to achieve

I'll send more info about this through discord :)